File: spring-restdocs/spring-restdocs-webtestclient/src/main/java/com/gmail/unmacaque/spring/config/RouteConfiguration.java
package com.gmail.unmacaque.spring.config;

import com.gmail.unmacaque.spring.domain.MessageRepository;
import com.gmail.unmacaque.spring.webflux.MessageRouteHandler;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration
public class RouteConfiguration {

	@Bean
	public MessageRouteHandler routeHandler(MessageRepository messageRepository) {
		return new MessageRouteHandler(messageRepository);
	}

	@Bean
	public RouterFunction<ServerResponse> routerFunction(MessageRouteHandler handler) {
		return route()
				.GET("/messages", handler::getMessages)
				.POST("/messages", handler::postMessage)
				.build();
	}
}

File: projects/kotlin/src/main/kotlin/com/gmail/unmacaque/spring/kotlin/config/RouteConfiguration.kt
package com.gmail.unmacaque.spring.kotlin.config

import com.gmail.unmacaque.spring.kotlin.domain.MessageRepository
import com.gmail.unmacaque.spring.kotlin.web.HelloHandler
import com.gmail.unmacaque.spring.kotlin.web.MessageHandler
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.servlet.function.router

@Configuration(proxyBeanMethods = false)
@EnableConfigurationProperties(ApplicationProperties::class)
class RouteConfiguration {

    @Bean
    fun helloHandler(properties: ApplicationProperties) =
        HelloHandler(properties.text)

    @Bean
    fun routeHandler(repository: MessageRepository) =
        MessageHandler(repository)

    @Bean
    fun routerFunction(helloHandler: HelloHandler, messageHandler: MessageHandler) =
        router {
            "/hello".nest {
                GET("/", helloHandler::sayHello)
            }
            "/messages".nest {
                GET("/", messageHandler::getMessages)
                GET("/{id}", messageHandler::getMessage)
                POST("/", messageHandler::postMessage)
            }
        }

}
